File: libs/go/faults/retry.go
// Copyright © 2026 Mindclade, LLC. All Rights Reserved.
// Mindclade Proprietary and Confidential.
// SPDX-License-Identifier: LicenseRef-Mindclade-Proprietary
//

package faults

import "time"

// RetryKind describes how a caller may retry an operation. It does not execute
// retries or prescribe a particular backoff algorithm.
type RetryKind string

const (
	RetryKindUnspecified RetryKind = ""
	RetryKindNever       RetryKind = "never"
	RetryKindImmediate   RetryKind = "immediate"
	RetryKindBackoff     RetryKind = "with_backoff"
	RetryKindAfter       RetryKind = "after"
)

// RetryPolicy communicates retry intent to a caller.
//
// MaxAttempts is the maximum total number of attempts, including the initial
// attempt. Zero delegates the limit to the caller's policy.
type RetryPolicy struct {
	Kind        RetryKind     `json:"kind,omitempty"`
	After       time.Duration `json:"after,omitempty"`
	MaxAttempts int           `json:"max_attempts,omitempty"`
}

// Valid reports whether the policy is already in canonical form.
func (policy RetryPolicy) Valid() bool {
	normalized := policy.Normalized()
	return policy == normalized
}

// Normalized returns a canonical and safe representation of the policy.
func (policy RetryPolicy) Normalized() RetryPolicy {
	if policy.MaxAttempts < 0 {
		policy.MaxAttempts = 0
	}

	switch policy.Kind {
	case RetryKindUnspecified:
		policy.After = 0
		policy.MaxAttempts = 0
	case RetryKindNever:
		policy.After = 0
		policy.MaxAttempts = 0
	case RetryKindImmediate:
		policy.After = 0
	case RetryKindBackoff:
		policy.After = 0
	case RetryKindAfter:
		if policy.After <= 0 {
			policy.Kind = RetryKindImmediate
			policy.After = 0
		}
	default:
		policy = RetryPolicy{}
	}

	return policy
}

// Specified reports whether the policy explicitly communicates retry intent.
func (policy RetryPolicy) Specified() bool {
	return policy.Normalized().Kind != RetryKindUnspecified
}

// Retryable reports whether the normalized policy permits another attempt.
func (policy RetryPolicy) Retryable() bool {
	normalized := policy.Normalized()
	if normalized.MaxAttempts == 1 {
		return false
	}

	switch normalized.Kind {
	case RetryKindImmediate, RetryKindBackoff, RetryKindAfter:
		return true
	default:
		return false
	}
}

// NoRetry returns an explicit policy that forbids retrying.
func NoRetry() RetryPolicy {
	return RetryPolicy{Kind: RetryKindNever}
}

// ImmediateRetry permits an immediate retry.
func ImmediateRetry(maxAttempts int) RetryPolicy {
	return RetryPolicy{
		Kind:        RetryKindImmediate,
		MaxAttempts: maxAttempts,
	}.Normalized()
}

// BackoffRetry permits retrying with a caller-selected backoff algorithm.
func BackoffRetry(maxAttempts int) RetryPolicy {
	return RetryPolicy{
		Kind:        RetryKindBackoff,
		MaxAttempts: maxAttempts,
	}.Normalized()
}

// DelayedRetry permits retrying after at least after has elapsed. A non-positive
// delay is normalized to an immediate retry.
func DelayedRetry(after time.Duration, maxAttempts int) RetryPolicy {
	return RetryPolicy{
		Kind:        RetryKindAfter,
		After:       after,
		MaxAttempts: maxAttempts,
	}.Normalized()
}

File: libs/python/errors/retry.py
# Copyright © 2026 Mindclade, LLC. All Rights Reserved.
# Mindclade Proprietary and Confidential.
# SPDX-License-Identifier: LicenseRef-Mindclade-Proprietary
#

"""Retry *intent*, carried alongside a fault.

This module communicates whether and how a caller may try again. It does not
retry anything, does not sleep, and does not prescribe a backoff curve — the same
split ``libs/go/faults`` and ``libs/go/retry`` keep, and for the same reason: the
side that knows why an operation failed is rarely the side that should own the
schedule for trying it again.

Durations are integer milliseconds rather than ``timedelta``. The value crosses a
process boundary into Go's ``time.Duration`` and Rust's ``Duration``, and a
float-backed ``timedelta`` round-trips through JSON with a precision the other two
languages do not share.
"""

from __future__ import annotations

from dataclasses import dataclass, replace
from enum import StrEnum


class RetryKind(StrEnum):
    """How a caller may retry.

    ``UNSPECIFIED`` is the empty string, matching Go's zero value, so a fault that
    never expressed retry intent and one that expressed "no opinion" serialize
    identically instead of becoming two states a consumer has to tell apart.
    """

    UNSPECIFIED = ""
    NEVER = "never"
    IMMEDIATE = "immediate"
    BACKOFF = "with_backoff"
    AFTER = "after"


@dataclass(frozen=True, slots=True)
class RetryPolicy:
    """Retry intent communicated to a caller.

    ``max_attempts`` counts the initial attempt, so ``1`` means "do not try again"
    and ``0`` delegates the limit to the caller's own policy.
    """

    kind: RetryKind = RetryKind.UNSPECIFIED
    after_millis: int = 0
    max_attempts: int = 0

    def normalized(self) -> RetryPolicy:
        """Return the canonical form of this policy.

        Every field combination that cannot mean anything is collapsed onto one
        that can: a delay on a kind that has no delay is dropped, and ``AFTER``
        with a non-positive delay becomes ``IMMEDIATE`` — it already behaves that
        way, and saying so makes two equivalent policies compare equal.
        """
        max_attempts = max(self.max_attempts, 0)

        match self.kind:
            case RetryKind.UNSPECIFIED | RetryKind.NEVER:
                return replace(self, after_millis=0, max_attempts=0)
            case RetryKind.IMMEDIATE | RetryKind.BACKOFF:
                return replace(self, after_millis=0, max_attempts=max_attempts)
            case RetryKind.AFTER:
                if self.after_millis <= 0:
                    return replace(
                        self,
                        kind=RetryKind.IMMEDIATE,
                        after_millis=0,
                        max_attempts=max_attempts,
                    )
                return replace(self, max_attempts=max_attempts)
            case _:
                return RetryPolicy()

    def valid(self) -> bool:
        """Report whether this policy is already canonical."""
        return self == self.normalized()

    def specified(self) -> bool:
        """Report whether the policy expresses retry intent at all."""
        return self.normalized().kind is not RetryKind.UNSPECIFIED

    def retryable(self) -> bool:
        """Report whether the normalized policy permits another attempt."""
        normalized = self.normalized()
        if normalized.max_attempts == 1:
            return False
        return normalized.kind in {
            RetryKind.IMMEDIATE,
            RetryKind.BACKOFF,
            RetryKind.AFTER,
        }


def no_retry() -> RetryPolicy:
    """An explicit policy forbidding another attempt."""
    return RetryPolicy(kind=RetryKind.NEVER)


def immediate_retry(max_attempts: int = 0) -> RetryPolicy:
    """Permit an immediate retry."""
    return RetryPolicy(kind=RetryKind.IMMEDIATE, max_attempts=max_attempts).normalized()


def backoff_retry(max_attempts: int = 0) -> RetryPolicy:
    """Permit retrying under a caller-selected backoff algorithm."""
    return RetryPolicy(kind=RetryKind.BACKOFF, max_attempts=max_attempts).normalized()


def delayed_retry(after_millis: int, max_attempts: int = 0) -> RetryPolicy:
    """Permit retrying after at least ``after_millis`` has elapsed."""
    return RetryPolicy(
        kind=RetryKind.AFTER, after_millis=after_millis, max_attempts=max_attempts
    ).normalized()
